hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_scan_response_data(const uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Get the current BLE scan response data.
+ *
+ * @param[in,out]   buf     Pointer to the scan response data to be filled.
+ * @param[in]       len     Length of the given buffer.
+ *
+ * @returns     The length of copied data.
+ */
+size_t ble_gap_get_scan_response_data(uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Start BLE advertising.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_start_advertising(void* reserved);
+
+/**
+ * Stop BLE advertising.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_stop_advertising(void);
+
+/**
+ * Check if BLE is advertising.
+ *
+ * @returns     true if it is advertising, otherwise, false.
+ */
+bool ble_gap_is_advertising(void);
+
+/**
+ * Set the BLE scanning parameters.
+ *
+ * @param[in]   scan_params Pointer to the scanning parameters to be set.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_scan_parameters(const hal_ble_scan_params_t* scan_params, void* reserved);
+
+/**
+ * Get the BLE scanning parameters.
+ *
+ * @param[in,out]   scan_params Pointer to the scanning parameters to be filled.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_scan_parameters(hal_ble_scan_params_t* scan_params, void* reserved);
+
+/**
+ * Start scanning nearby BLE devices.
+ *
+ * @param[in] callback  The callback function to handle the scan results.
+ * @param[in] context   The callback function context.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_start_scan(on_ble_scan_result_cb_t callback, void* context, void* reserved);
+
+/**
+ * Check if BLE is scanning nearby devices.
+ *
+ * @returns     true if it is scanning, otherwise false.
+ */
+bool ble_gap_is_scanning(void);
+
+/**
+ * Stop scanning nearby BLE devices.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_stop_scan(void);
+
+/**
+ * Connect to a peer BLE device.
+ *
+ * @param[in]   address Pointer to the peer BLE identity address.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_connect(const hal_ble_addr_t* address, void* reserved);
+
+/**
+ * Check if BLE is connecting with peer device.
+ *
+ * @returns true if connecting, otherwise false.
+ */
+bool ble_gap_is_connecting(void);
+
+/**
+ * Check if BLE is connected with peer device.
+ *
+ * @param[in]   address Pointer to the peer BLE identity address.
+ *                      If NULL, it returns true as long as there has on-going connections.
+ *
+ * @returns true if connected, otherwise false.
+ */
+bool ble_gap_is_connected(const hal_ble_addr_t* address);
+
+/**
+ * Cancel the ongoing procedure that connecting to the peer BLE device.
+ *
+ * @returns true if success, otherwise false.
+ */
+int ble_gap_connect_cancel(void);
```

Comment:
`ble_gap_connect()` and `ble_gap_is_connected()` both have an `address` parameter. It should probably be added here is as well.